### PR TITLE
Cgt 1908

### PR DESCRIPTION
**Pull request by euptonqa**
Contains the Who Did You Give The Property to page along with testing.

### CODE REVIEW

File: app/controllers/resident/properties/GainController.scala [hunk 1]
```diff
@@ -128,9 +129,35 @@ trait GainController extends FeatureLock {
     )
   }
 
+  val youHaveNoTaxToPay = FeatureLockForRTT.async { implicit request =>
+    TODO.apply(request)
+  }
+
+  val whatWasItWorthWhenYouGaveItAway = FeatureLockForRTT.async { implicit request =>
```
Comment (jameshforster) [inline app/controllers/resident/properties/GainController.scala]: No longer needed, method exists in controller.

File: app/forms/resident/properties/gain/PropertyRecipientForm.scala [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import models.resident.properties.gain.PropertyRecipientModel
+import play.api.data.Forms._
+import play.api.data._
+import play.api.i18n.Messages
+
+object PropertyRecipientForm {
+  val propertyRecipientForm = Form(
+    mapping("propertyRecipient" -> text
+        .verifying(Messages("calc.resident.whoDidYouGiveThePropertyTo.error.mandatory"), mandatoryCheck)
+        .verifying(Messages("calc.resident.whoDidYouGiveThePropertyTo.error"), whoDidYouGiveItTooCheck)
```
Comment (jameshforster) [inline app/forms/resident/properties/gain/PropertyRecipientForm.scala]: Should be same error as validation above.

File: conf/messages [hunk 3]
```diff
@@ -223,6 +223,14 @@ calc.privateResidenceRelief.error.errorDecimalPlaces = There are too many number
 calc.privateResidenceRelief.error.maxNumericExceeded = Enter a value for your days claimed that''s
 calc.privateResidenceRelief.error.maxNumericExceeded.OrLess = or less
 
+## Who did you give the property to ##
+calc.resident.whoDidYouGiveThePropertyTo.title = Who did you give the property to?
+calc.resident.whoDidYouGiveThePropertyTo.spouse = Your spouse or a civil partner
+calc.resident.whoDidYouGiveThePropertyTo.charity = A charity
+calc.resident.whoDidYouGiveThePropertyTo.other = Someone else
+calc.resident.whoDidYouGiveThePropertyTo.error = Please select a valid option
```
Comment (jameshforster) [inline conf/messages]: Not a required error

File: test/assets/MessageLookup.scala [hunk 4]
```diff
@@ -438,4 +438,12 @@ object MessageLookup {
     val gift = "Gave it away"
   }
 
+  object whoDidYouGiveItTo {
+    val title = "Who did you give the property to?"
+    val spouse = "Your spouse or a civil partner"
+    val charity = "A charity"
+    val other = "Someone else"
+    val error = "Please select a valid option"
```
Comment (jameshforster) [inline test/assets/MessageLookup.scala]: Not a required error

File: test/controllers/resident/properties/GainControllerSpec/PropertyRecipientActionSpec.scala [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+      "redirect" in {
+        status(result) shouldEqual 303
+      }
+
+      "to the  page" in {
+        redirectLocation(result) shouldBe Some("/calculate-your-capital-gains/resident/properties/you-have-no-tax-to-pay")
+      }
+    }
+  }
+
```
Comment (jameshforster) [inline test/controllers/resident/properties/GainControllerSpec/PropertyRecipientActionSpec.scala]: Please add seperate test for instance of Spouse please and one for a bad request check

File: test/routes/properties/RoutesSpec.scala [hunk 6]
```diff
@@ -59,6 +59,18 @@ class RoutesSpec extends UnitSpec with WithFakeApplication with Matchers {
...
 
+  "The URL for GET for the Who Did You Give It To action" should {
+    "be equal to /calculate-your-capital/gains/resident/properties/who-did-you-give-it-to" in {
+      GainController.whoDidYouGiveItTo().url shouldEqual "/calculate-your-capital-gains/resident/properties/who-did-you-give-it-to"
+    }
+  }
+
+  "The URL for POST for the Who Did You Give It To action" should {
+    "be equal to /calculate-your-capital/gains/resident/properties/who-did-you-give-it-to" in {
+      GainController.whoDidYouGiveItTo().url shouldEqual "/calculate-your-capital-gains/resident/properties/who-did-you-give-it-to"
```
Comment (jameshforster) [inline test/routes/properties/RoutesSpec.scala]: This is calling the GET request, should call submitWhoDidYouGiveItTo

File: test/views/resident/properties/gain/PropertyRecipientViewSpec.scala [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        form.attr("action") shouldBe controllers.resident.properties.routes.GainController.submitWhoDidYouGiveItTo().toString
+      }
+
+      "has the method of POST" in {
+        form.attr("method") shouldBe "POST"
+        }
+
+    }
+
+    "have additional content that" should {
```
Comment (jameshforster) [inline test/views/resident/properties/gain/PropertyRecipientViewSpec.scala]: Wording of this is confusing (not additional content or bullet points, it refers to inputs)

Refer to how-became-owner viewspec for examples

File: test/routes/properties/RoutesSpec.scala (hunk 6 above)
Comment (euptonqa) [inline test/routes/properties/RoutesSpec.scala]: It goes to two different URLs depending on the data entered in the form, not entirely sure how to handle this.

Comment (DavidWoodward37) [inline test/routes/properties/RoutesSpec.scala]: Ah but how the POST works is when you call the method it POST's to the URL specified in the route :) (in this case it posts to /calculate-your-capital-gains/resident/properties/who-did-you-give-it-to, which is the same route as the get action.)

**Comment by DavidWoodward37** [comment]
Merge conflicts too sorry!

**Comment by DavidWoodward37** [comment]
Build failed due to timed out futures, just restarting it now.

**Comment by DavidWoodward37** [comment]
More merge conflicts to fix sorry.

**Pull request merged**